File: frontend/src/types/portfolio.ts
// ── 포트폴리오 관련 타입 ─────────────────────────────────────────────────────

export type CurrencyCode = 'KRW' | 'USD' | 'JPY' | 'EUR' | 'HKD'
export type AssetClass   = 'STOCK' | 'ETF' | 'BOND' | 'CRYPTO' | 'CASH'
export type MarketType   = 'KRX' | 'NASDAQ' | 'NYSE' | 'AMEX' | 'OTHER'
export type TransactionType = 'BUY' | 'SELL' | 'DIVIDEND'

// ── Portfolio ─────────────────────────────────────────────────────────────

export interface Portfolio {
  id: string
  name: string
  description: string | null
  base_currency: CurrencyCode
  is_active: boolean
  created_at: string
}

export interface PortfolioCreate {
  name: string
  description?: string
  base_currency: CurrencyCode
}

// ── Holding ───────────────────────────────────────────────────────────────

export interface Holding {
  id: string
  portfolio_id: string
  ticker: string
  asset_class: AssetClass
  market: MarketType
  quantity: string       // Decimal → string
  average_cost: string
  currency_code: CurrencyCode
  created_at: string
}

export interface HoldingCreate {
  ticker: string
  asset_class: AssetClass
  market: MarketType
  quantity: string
  average_cost: string
  currency_code: CurrencyCode
  execution_exchange_rate: string
}

// ── Valuation (현재가 반영 수익률) ─────────────────────────────────────────

export interface HoldingValuation {
  ticker: string
  quantity: number
  average_cost: number
  current_price: number
  currency_code: CurrencyCode
  current_price_krw: number
  cost_krw: number
  gain_loss_krw: number
  gain_loss_pct: number
  price_contribution_krw: number
  fx_contribution_krw: number
}

export interface PortfolioValuation {
  portfolio_id: string
  portfolio_name: string
  base_currency: CurrencyCode
  usd_krw_rate: number
  rate_source: string
  total_cost_krw: number
  total_value_krw: number
  total_return_pct: number
  price_contribution_krw: number
  fx_contribution_krw: number
  holdings: HoldingValuation[]
}

File: app/schemas/portfolio.py
"""
Portfolio / Holding 관련 Pydantic 스키마

핵심 제약: 이 플랫폼은 실제 매수/매도 실행 기능이 없음.
Holdings는 사용자가 '직접 수동 입력'하는 보유 내역 기록임.
"""

import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field, field_validator

from app.core.enums import AssetClass, CurrencyCode, MarketType


# ─────────────────────────────────────────────────────────────────────────────
# Portfolio
# ─────────────────────────────────────────────────────────────────────────────

class PortfolioCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    description: str | None = Field(default=None, max_length=500)
    base_currency: CurrencyCode = CurrencyCode.KRW


class PortfolioUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    description: str | None = None
    base_currency: CurrencyCode | None = None


class PortfolioResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    owner_id: uuid.UUID
    name: str
    description: str | None
    base_currency: str
    is_active: bool
    created_at: datetime
    updated_at: datetime


# ─────────────────────────────────────────────────────────────────────────────
# Holding  (수동 입력 보유 종목)
# ─────────────────────────────────────────────────────────────────────────────

class HoldingCreate(BaseModel):
    """
    사용자가 직접 입력하는 보유 종목 등록 스키마.

    주의: execution_exchange_rate는 매수 당시 환율을 사용자가 직접 기입.
          KRW 자산이면 1.0 으로 입력.
    """

    ticker: str = Field(
        min_length=1,
        max_length=20,
        description="거래소 티커 (예: 005930.KS 또는 AAPL)",
    )
    asset_class: AssetClass
    market: MarketType
    quantity: Decimal = Field(
        gt=Decimal("0"),
        description="보유 수량 (소수 단위 지원)",
    )
    average_cost: Decimal = Field(
        gt=Decimal("0"),
        description="평균 매수 단가 (currency_code 기준)",
    )
    currency_code: CurrencyCode
    execution_exchange_rate: Decimal = Field(
        gt=Decimal("0"),
        description="매수 당시 환율 (1 USD = N KRW). KRW 자산은 1.0",
    )

    @field_validator("ticker")
    @classmethod
    def uppercase_ticker(cls, v: str) -> str:
        return v.upper().strip()


class HoldingUpdate(BaseModel):
    """수량 또는 평균 단가 수정 (종목 자체는 변경 불가 — 삭제 후 재등록)"""

    quantity: Decimal | None = Field(default=None, gt=Decimal("0"))
    average_cost: Decimal | None = Field(default=None, gt=Decimal("0"))
    execution_exchange_rate: Decimal | None = Field(default=None, gt=Decimal("0"))


class HoldingResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    portfolio_id: uuid.UUID
    ticker: str
    asset_class: AssetClass
    market: MarketType
    quantity: Decimal
    average_cost: Decimal
    currency_code: CurrencyCode
    created_at: datetime
    updated_at: datetime
